plugins/inputs/kafka_topic_consumer/kafka_topic_consumer.go

```diff
@@ -0,0 +1,290 @@
+	var err error
+	var wg = sync.WaitGroup{}
+
+	topics, err := k.client.Topics()
+	if err != nil {
+		k.Log.Warn("fetch topics info error: %s", err)
+		return err
+	}
+
+	topicPartitionsOffsets := make(map[string]map[int32]int64)
+	for _, topic := range topics {
+		if topic == "" || !k.topicFilter.MatchString(topic) || (k.excludeTopics != nil && k.excludeTopics.MatchString(topic)) {
+			continue
+		}
+		topicPartitionsOffsets[topic], err = k.fetchTopicPartitionOffset(topic)
+		if err != nil {
+			k.Log.Warnf("fetch topic: %s, partition infomation error: %s", topic, err)
+			continue
+		}
+	}
+
+	brokers := k.client.Brokers()
+	if len(brokers) == 0 {
+		k.Log.Warnf("the kafka has't any brokers")
+		return nil
+	}
+	groupLags := make([]groupLag, 0)
+	for _, broker := range brokers {
+		wg.Add(1)
+		go func(broker *sarama.Broker) {
+			defer wg.Done()
+			if err := broker.Open(k.client.Config()); err != nil && err != sarama.ErrAlreadyConnected {
+				k.Log.Errorf("cannot connect to broker %d: %v", broker.ID(), err)
+				return
+			}
+
+			groupLag := k.fetchLagData(broker, topicPartitionsOffsets)
+			groupLags = append(groupLags, groupLag...)
+		}(broker)
+	}
+	wg.Wait()
+
+	k.writeToAccumulator(groupLags, topicPartitionsOffsets, acc)
+	return nil
+}
+
+func (k *kafkaTopicConsumer) fetchTopicPartitionOffset(topic string) (map[int32]int64, error) {
+	partitions, err := k.client.Partitions(topic)
+	if err != nil {
+		return nil, err
+	}
+	result := make(map[int32]int64, len(partitions))
+	for _, partition := range partitions {
+		currentOffset, err := k.client.GetOffset(topic, partition, sarama.OffsetNewest)
+		if err != nil {
+			k.Log.Warnf("fetch topic: %s, partition: %d newest offset error: %s", topic, partition, err)
+			continue
+		}
+		result[partition] = currentOffset
+	}
+	return result, nil
+
+}
+
+func (k *kafkaTopicConsumer) fetchLagData(broker *sarama.Broker, topicPatitionsOffsets map[string]map[int32]int64) []groupLag {
+	groups, err := broker.ListGroups(&sarama.ListGroupsRequest{})
+	if err != nil {
+		k.Log.Errorf("cannot get consumer group: %v", err)
+		return nil
+	}
+	groupIds := make([]string, 0)
+	for groupId := range groups.Groups {
+		if k.groupFilter.MatchString(groupId) {
+			groupIds = append(groupIds, groupId)
+		}
+	}
+
+	result := make([]groupLag, 0)
+	for _, group := range groupIds {
+		offsetFetchResponse := k.getGroupAllTopicOffset(broker, group, topicPatitionsOffsets)
+		for topic, partitions := range offsetFetchResponse.Blocks {
+			var currentOffsetSum int64
+			var lagSum int64
+			partitionConsumed := false
+			items := make([]groupLagItem, 0)
+			for partition, offsetResponseBlock := range partitions {
+				err := offsetResponseBlock.Err
+				if err != sarama.ErrNoError {
+					k.Log.Errorf("partition %d offset of the topic: %s and consumer group: %s error :%v", partition, topic, group, err.Error())
+					continue
+				}
+				currentOffset := offsetResponseBlock.Offset
+				currentOffsetSum += currentOffset
```

Comment:
Do we need to check the case where `currentOffset` is `-1`

Reply:
Updated